### DISCUSSION TITLE
RFC: Zarr-Python 3.0 Design Doc

### DISCUSSION THREAD

**Discussion by jhamman** [zarr-developers/zarr-python#1569] [Announcements]
As part of the Zarr-Python Refactoring working group (https://github.com/zarr-developers/zarr-python/discussions/1480) I have developed a [draft proposal](https://hackmd.io/@nb-2zB9iQQe2vOmdhVedkQ/SyyPYe_76/edit) for a major block of work in the Zarr-Python project, leading version 3.0 and including support for the V3 spec. This post is both a _Request for Comment_ and a forum for discussion. 

### Goals (read the doc for details)
- Provide a complete implementation of Zarr V3 through the Zarr-Python API
- Align the Zarr-Python array API with the [array API Standard](https://data-apis.org/array-api/latest/)
- Clear the way for exciting extensions / ZEPs (i.e. sharding, variable chunking, etc.)
- Provide a developer API that can be used to implement and register V3 extensions
- Improve the performance of Zarr-Python by streamlining the interface between the Store layer and higher level APIs (e.g. Groups and Arrays)
- Clean up the internal and user facing APIs 
- Improve code quality and robustness (e.g. achieve 100% type hint coverage)

Thanks to @d-v-b, @normanrz, @rabernat, @dcherian, @monodeldiablo, @olimcc, @martindurant, and @JackKelly for their input so far. cc @zarr-developers/python-core-devs.

Note: The next Zarr-Python Refactor working group is slated to meet at 9a PT on November 22, 2023. We'll be discussing this topic in detail then, please feel free to join (https://github.com/zarr-developers/zarr-python/discussions/1480).

**Comment by JackKelly** [comment]
Looks awesome!

## Working with _batches_ of chunks when doing IO and/or (de)compression from sharded Zarrs

**TL;DR:** The current design looks great for handling something like 1,000 chunks per second[^1]. But, if we want to push Zarr to handle more like one million chunks per second (from a sharded Zarr), then it'd be great to discuss adding support for processing _batches_ of chunks in one go.

I'd be keen to discuss whether this draft proposal would be an appropriate time to consider adding support for processing _batches_ of chunks (in a single function call). For example, allowing `Array` to ask the lower layers to load & decompress a million chunks, spread across a small number of shards, all in a single function call. There are several possible advantages, most of which have been discussed before:

* Batched IO would allow `Store`s to apply optimisations such as:
  - Merging multiple nearby reads (as proposed by @bilts in #547, back in 2020)
  - Exploit parallelism in some storage subsystems (e.g. SSDs and cloud storage buckets)
  - Exploit batched APIs exposed by the storage subsystem (e.g. [vectored IO](https://en.wikipedia.org/wiki/Vectored_I/O)[^2] like [POSIX `readv()`](https://pubs.opengroup.org/onlinepubs/9699919799/) and [Windows `ReadFileScatter()`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-readfilescatter); and [multipart HTTP range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests#multipart_ranges), although [cloud storage APIs don't support multipart range requests](https://github.com/apache/arrow-rs/issues/4612#issuecomment-1661799951)) 
  - Allow the `Store` to plan properly: to pre-allocate buffers, spin up multiple threads, and distribute the workload across threads without using locks.
* Batched (de)compression would make it easier to (de)compress in parallel across multiple CPUs cores or on a GPU (as proposed by @akshaysubr in #1398)

Yes, in principle, the `async get(key: str) -> bytes:` function proposed in [the `Store` API in the current version of the Zarr-Python 3.0 doc](https://hackmd.io/0DVKP6d9QI-VaHc0zvOuxw#Store-API) does allow for parallel IO (by calling `store.get(key2)` before `get(key1)` has completed). But this makes it harder (or, in some cases, impossible) for the `Store` implementation to apply relevant optimisations[^3].

And, if we're trying to load one million chunks per second then we'll bump into the issue that function calls are surprisingly slow in Python: [each function call (with arguments) takes about 350 ns](https://ilovesymposia.com/2015/12/10/the-cost-of-a-python-function-call/)[^4] (_just_ for the function call: not doing any actual work!) A million function calls would take 0.35 seconds! So, it should be more efficient to do single `get_items` function call, with a million chunks as the function argument.

### Some reasons why we might want to ignore this suggestion for Zarr-Python:

* Maybe Zarr-Python isn't the right place to try to implement "extreme Zarr performance"? Pure-Python code will struggle to max out the hardware (for IO). Although it might be interesting to consider a world in which a Zarr "front-end" is implemented in Python, but some `Store`s are implemented in a compiled language (with a Python API).
* Maybe I'm the only one who cares about (or thinks it's sane to consider!) loading one million chunks per second?!?
* Perhaps adding support for _batched_ processing should be a topic of discussion for a future update to the Zarr spec, not this update to Zarr-Python.

### Footnotes

[^1]: I've picked the number "1,000" from thin air, TBH!
[^2]: Although I haven't been able to find any benchmarks to show that vectored I/O is actually _faster_, especially when using Linux io_uring or Windows I/O Ring. (Because io_uring and I/O Ring allow for multiple `read` requests to be submitted in a single system call. Which, perhaps, negates the obvious advantage to vectored I/O)
[^3]: `Store` could collect multiple `get()` requests by waiting a while, to see if it gets any more `get()` requests, before performing any IO. But requiring the `Store` to wait feels like it defeats the objective of going fast! And it adds complexity, and the time overhead of performing a million function calls in Python.
[^4]: The figure of 350 ns is from an old (2015) blog post. Since then, Python 3.10 and 3.11 sped up function calls a bit (20%??). But I couldn't find any more modern benchmarks.

**Reply by normanrz** [reply]
In zarrita it would be trivial to add a batch-mode, because reading already goes through a compact loop https://github.com/scalableminds/zarrita/blob/async/zarrita/array.py#L302-L309

Looking at the zarr-python proposal, I feel this would need another abstraction that handles the fetching, decompression, array assembly. Fwiw, the proposal already contains an API for batching requests within a single shard with `async def get_partial_values(self, key_ranges: List[Tuple[str, int]]) -> bytes:`.

**Reply by jhamman** [reply]
Thanks @JackKelly for bringing this up. It's an important use case and one that I agree is currently not entirely handled in the design doc. I'm open to reworking the Store interface or to add another abstraction that sits between the Array and Store. As @normanrz points out, we have `get_partial_values`. To me, the question we need to answer is where this would be best solved, within the Store or higher up in the stack?

**Reply by normanrz** [reply]
I think it would have to be higher up than the store, if it should also handle codecs. I made a quick prototype in zarrita (just the read path): https://github.com/scalableminds/zarrita/compare/chunk-fetcher
Batching could be useful not only for reading and decoding, but also for indexing. At that point, it almost feels like a separate `Array` implementation.

**Reply by jni** [reply]
Ha @JackKelly it is very gratifying to have someone cite my own blog post back at me. 😂 👋 

> The figure of 350 ns is from an old (2015) blog post. Since then, Python 3.10 and 3.11 sped up function calls.

This prompted me to revisit things. On my current machine on Python 3.12, it's now just under 50ms for 1M function calls or 50ns per call:

```
Python 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:36:57) [Clang 15.0.7 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.17.2 -- An enhanced Interactive Python. Type '?' for help.

In [2]: def f():
   ...:     pass
   ...:
In [3]: def g(n):
   ...:     for i in range(n):
   ...:         pass
   ...:
In [4]: def h(n):
   ...:     for i in range(n):
   ...:         f()
   ...:
In [6]: n = int(1e6)

In [7]: %timeit g(n)
12.2 ms ± 21.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [8]: %timeit h(n)
28.6 ms ± 595 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [9]: def f(*args, **kwargs):
   ...:     pass
   ...:

In [10]: def h(n):
    ...:     for i in range(n):
    ...:         f(n)
    ...:

In [11]: %timeit h(n)
61.6 ms ± 3.62 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Having said this, it seems nearly all of the gains are from the processor, *not* from CPython improvements, at least not since 3.8, which is the earliest version I can install from conda-forge:

```
Python 3.8.18 | packaged by conda-forge | (default, Oct 10 2023, 15:46:56)
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.2 -- An enhanced Interactive Python. Type '?' for help.

In [1]: def f(*args, **kwargs):
   ...:     pass
   ...:

In [2]: def g(n):
   ...:     for i in range(n):
   ...:         pass
   ...:

In [3]: def h(n):
   ...:     for i in range(n):
   ...:         f(n)
   ...:

In [4]: n = int(1e6)

In [5]: %timeit g(n)
10.9 ms ± 418 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [6]: %timeit h(n)
62 ms ± 2.35 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

50ns is a lot better than 350ns, but still could be a significant chunk of time when reading a chunk (call it 1MB), especially if you have a couple of levels of indirection from the user-facing API, which seems likely. So +1 to ensuring we have a batch strategy, one way or another. (I don't have an opinion about which level it should live on — I have only interacted with zarr using public APIs 😅.) As a user I would definitely like zarr-python to be competitive in performance, even if it's not the top — I just wouldn't want it to be 10-100x slower than the top. 🐢 😂 

On the other hand, for decent-size chunks, you can get [very good throughput](https://fosstodon.org/@jni/109902906137306269) using pure python (4GB/s), so maybe this is not the top priority and could be discussed later...

**Reply by martindurant** [reply]
I would say it's not really the loop or the function call that will make the difference, but (as other benchmarking already found) memcopies, cpython API calls and syscalls. For instance
```
def g(n):
    for i in range(n):
        pass
```
10.2ms

```
def g(n):
    for i in range(n):
        b[10:20]
```
(where b is a bytes object)
61ms

**Reply by JackKelly** [reply]
Hi @jni, ha! Sorry I didn't "connect the dots" between your blog and your GitHub ID! It's a small world, huh?!? Thanks loads for re-doing the benchmarks! 

@martindurant, I completely agree that memcopies are surprisingly slow. As you've mentioned in the past, the ideal case would be that - for uncompressed Zarr chunks - we can do exactly one copy: from storage to the final numpy array (ideally using DMA). If we have batched processing of chunks, then there's at least a chance that, even for chunks that need to be "scattered" throughout the final numpy array, it'd still be possible to do that using DMA (e.g. using [`readv`](https://man7.org/linux/man-pages/man2/readv.2.html) to read different parts of a single chunk into different memory locations, with a single system call).

**Reply by martindurant** [reply]
I was not really advocating for a low level super fix like that, but by all means try it. I think the python readinto and decompress_into are probably enough, but the point is that the "batch" concept I don't think will make much difference so long as high latency reads are awaited together. I consider this the low-hanging fruit, together with not reading bytes we don't need. 

All contingent on contiguous memory blocks. Where striding is at issue, we probably can't do better than numpy copies, unless we write our own decompression wrapper capable of "getting every fourth uncompressed byte" or similar pattern.

**Reply by JackKelly** [reply]
On the topic of where to put this in the stack...

I feel I'm too new to the Zarr-Python world to have a particularly strong grasp of the pros and cons of where to place this. But I will give it more thought.

One option that [I'm considering](https://github.com/JackKelly/light-speed-io/pull/7/files#diff-6d5ca28a7b2d3b2e4cc33389e06c30488c1f9d618a792865720ac8b3fe54b165R51-R56) for my Zarr Rust experiments, is to have a method which - ported to Python - would be something like:

 ```python
class Store:
    async def get_items(
        self,
        keys: Iterable[str],
        transform: Optional[Callable] = None
    ) -> list[bytes]:
    """Get multiple chunks.

    Args:
        keys: List of chunk keys.
        transform: A function that will be applied to every chunk. The function must take two arguments:
                - chunk_id: int
                - chunk_data: bytes
            And it must return bytes or None (if, for example, the processing function moves data to a final array)
    """
```

Then the `Array` can ask the `Store` to perform any arbitrary transform on each chunk (e.g. decompress and copy to the final numpy array). My motivation is twofold: I want processing to start as soon as each chunk is available, using a threadpool (one thread per CPU core). And I want _all_ processing (decompression, maybe some numerical transform, and copying to the final array) to happen in very quick succession, to maximise that chance that the chunk stays in CPU cache.

So, this kind of dodges the question of "where to put decompression". The `Store` can perform any arbitrary computation. But the decompression algorithm(s) wouldn't by defined in the `Store`.

But I don't know if that's appropriate for Zarr-Python. Nor have I actually implemented this in Rust yet, so I don't know if it's a good idea for Rust, either!

**Reply by martindurant** [reply]
Were you following `rfsspec`? That does the concurrent reads; since cramjam is rust-python, I thought it could nicely "decompress into" numpy buffers in CPU threads driven by tokio.

**Comment by alimanfoo** [comment]
Hi all, just wanted to offer support for the roadmap and design doc (#1583), kudos for plotting a clear course through some very tricky terrain.
